api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java

```diff
@@ -691,9 +713,91 @@ public String getSecretQuestion(User user) throws APIException {
 		if (user.getUserId() != null) {
 			LoginCredential loginCredential = dao.getLoginCredential(user);
 			return loginCredential.getSecretQuestion();
-		} else {
+		} 
+		else {
 			return null;
 		}
     }
 	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByUsernameOrEmail(java.lang.String)
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByUsernameOrEmail(String usernameOrEmail) {
+		if (StringUtils.isNotBlank(usernameOrEmail)) {
+			User user = dao.getUserByEmail(usernameOrEmail);
+			if (user == null) {
+				return getUserByUsername(usernameOrEmail);
+			}
+			return user;
+		}
+		throw new APIException("error.usernameOrEmail.notNullOrBlank", (Object[]) null);
+	}
+	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByActivationKey(java.lang.String)
+	 * 
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByActivationKey(String activationKey) {
+		LoginCredential loginCred = dao.getLoginCredentialByActivationKey(activationKey);
+		if (loginCred != null) {
+			String[] credTokens = loginCred.getActivationKey().split(":");
```

Comment:
what happens if the `getActivationKey()` return null for an instance?

Reply:
Trusting the code that generation for the activation key will always be provided ie  setUserActivationKey(User user)  as that is it job.  and for this line of code to be executed the loginCred is not null which was gotten using the activation key hence we are guaranteed that if loginCred exist and is not null then activation key exist else null will be returned